### ISSUE TITLE
[QA] Regression: AV-Failed error with clamav

### ISSUE THREAD

**Issue by jnweiger** [owncloud/files_antivirus#491]
Seen while testing files_antivirus 1.1.0-RC with owncloud 10.10.0 on Ubuntu 22.04 and clamav engine 0.103.6

## Setup
* `clamscan eicar.com.txt` reports `/root/eicar.com.txt: Win.Test.EICAR_HDB-1 FOUND` OK.
* clamav is running in docker deepdiver/icap-clamav-service:latest (2020-11-03T15:26:15.943257956Z)
* Configuration:
![image](https://user-images.githubusercontent.com/1108546/176168785-bbfe4642-84a0-45d1-b2b4-0ff7c496ea4d.png)

## Steps to reproduce
* When uploading a (no-infected) photo via Web-UI, an unspecific AV Failed error occurs:
![image](https://user-images.githubusercontent.com/1108546/176169044-0e43a304-f032-4ab2-9686-3433821ac95f.png)
----
* The owncloud.log contains
```
{"reqId":"YrrlBBIpSkGUw1JzzjArJgAAAAM","level":0,"time":"2022-06-28T11:24:52+00:00","remoteAddr":"84.180.178.198","user":"admin","app":"files_antivirus
","method":"PUT","url":"\/remote.php\/dav\/files\/admin\/cybersecurity.png","message":"File size is 1425659. av_max_file_size is -1. Scanning is schedu
led."}
{"reqId":"YrrlBBIpSkGUw1JzzjArJgAAAAM","level":2,"time":"2022-06-28T11:24:59+00:00","remoteAddr":"84.180.178.198","user":"admin","app":"core","method":
"PUT","url":"\/remote.php\/dav\/files\/admin\/cybersecurity.png","message":"ignoring lock release with type 1 for files\/1b1515f57951c617176088ca17e65eec. Lock hasn't been acquired before"}
{"reqId":"YrrlBBIpSkGUw1JzzjArJgAAAAM","level":4,"time":"2022-06-28T11:24:59+00:00","remoteAddr":"84.180.178.198","user":"admin","app":"webdav","method":"PUT","url":"\/remote.php\/dav\/files\/admin\/cybersecurity.png","message":"Exception: HTTP\/1.1 500 AV failed!: {\"Exception\":\"Sabre\\\\DAV\\\\Exception\",\"Message\":\"AV failed!\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/File.php(243): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\File->convertToSabreException()\\n#1 \\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Directory.php(173): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\File->put()\\n#2 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(1098): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory->createFile()\\n#3 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(504): Sabre\\\\DAV\\\\Server->createFile()\\n#4 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/WildcardEmitterTrait.php(89): Sabre\\\\DAV\\\\CorePlugin->httpPut()\\n#5 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(472): Sabre\\\\DAV\\\\Server->emit()\\n#6 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(253): Sabre\\\\DAV\\\\Server->invokeMethod()\\n#7 \\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Server.php(345): Sabre\\\\DAV\\\\Server->start()\\n#8 \\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#9 \\\/var\\\/www\\\/owncloud\\\/remote.php(165): require_once('\\\/var\\\/www\\\/ownclo...')\\n#10 {main}\",\"File\":\"\\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/File.php\",\"Line\":702}"}
{"reqId":"YrrlBBIpSkGUw1JzzjArJgAAAAM","level":4,"time":"2022-06-28T11:24:59+00:00","remoteAddr":"84.180.178.198","user":"admin","app":"webdav","method":"PUT","url":"\/remote.php\/dav\/files\/admin\/cybersecurity.png","message":"Caused by: {\"Exception\":\"RuntimeException\",\"Message\":\"AV failed!\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/owncloud\\\/apps-external\\\/files_antivirus\\\/lib\\\/AvirWrapper.php(156): OCA\\\\Files_Antivirus\\\\Scanner\\\\ICAPScanner->completeAsyncScan()\\n#1 \\\/var\\\/www\\\/owncloud\\\/apps-external\\\/files_antivirus\\\/lib\\\/AvirWrapper.php(133): OCA\\\\Files_Antivirus\\\\AvirWrapper->onScanComplete()\\n#2 [internal function]: OCA\\\\Files_Antivirus\\\\AvirWrapper->OCA\\\\Files_Antivirus\\\\{closure}(*** sensitive parameters replaced ***)\\n#3 \\\/var\\\/www\\\/owncloud\\\/apps\\\/files_external\\\/3rdparty\\\/icewind\\\/streams\\\/src\\\/CallbackWrapper.php(119): call_user_func()\\n#4 [internal function]: Icewind\\\\Streams\\\\CallbackWrapper->stream_close()\\n#5 \\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/File.php(209): fclose()\\n#6 \\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Directory.php(173): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\File->put()\\n#7 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(1098): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory->createFile()\\n#8 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(504): Sabre\\\\DAV\\\\Server->createFile()\\n#9 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/WildcardEmitterTrait.php(89): Sabre\\\\DAV\\\\CorePlugin->httpPut()\\n#10 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(472): Sabre\\\\DAV\\\\Server->emit()\\n#11 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(253): Sabre\\\\DAV\\\\Server->invokeMethod()\\n#12 \\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Server.php(345): Sabre\\\\DAV\\\\Server->start()\\n#13 \\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#14 \\\/var\\\/www\\\/owncloud\\\/remote.php(165): require_once('\\\/var\\\/www\\\/ownclo...')\\n#15 {main}\",\"File\":\"\\\/var\\\/www\\\/owncloud\\\/apps-external\\\/files_antivirus\\\/lib\\\/Scanner\\\/ICAPScanner.php\",\"Line\":71}"}
{"reqId":"YrrlBBIpSkGUw1JzzjArJgAAAAM","level":1,"time":"2022-06-28T11:24:59+00:00","remoteAddr":"84.180.178.198","user":"admin","app":"core","method":"PUT","url":"\/remote.php\/dav\/files\/admin\/cybersecurity.png","message":"cleaning stray exclusive locks for files\/1b1515f57951c617176088ca17e65eec"}
{"reqId":"YrrlC1GagcIqumSmVKJsxgAAAAg","level":0,"time":"2022-06-28T11:24:59+00:00","remoteAddr":"84.180.178.198","user":"admin","app":"OC\\User\\Session::validateToken","method":"GET","url":"\/index.php\/apps\/files\/ajax\/getstoragestats.php?dir=%2F","message":"token 16ca312a80f3e2fe4740675c5e09bafc02d24f93a30555edfdd8f4843fad5b41a4af7349337ec4d8744c73dfb14718fa11c9d3ffc7f7b7f5de056cb27c2b59ff with token id 1 found, validating"}
```

## Epected behaviour:
* uninfected files upload fine.
* infected files are blocked with e.g.
![image](https://user-images.githubusercontent.com/1108546/176171350-eaf5b64f-a652-4988-b66d-da0d8e3dd0f3.png)

**Comment by jnweiger** [comment]
Identical behaviour with oc 10.9.1 on Ubuntu 20.04: the released files_antivius-1.0.0 works fine, the new 1.1.0-rc1 fails.

**Comment by jnweiger** [comment]
@dpapac @GeraldLeikam  this is the report from our session today.

**Comment by jvillafanez** [comment]
Works for me using the same docker image for icap.

Code seems to break in https://github.com/owncloud/files_antivirus/blob/2285d0ba5b9a44225827cf2d726130383925f7a1/lib/Scanner/ICAPScanner.php#L71 , which suggest a wrong response from the icap service. We might need either logs from the icap service or a wireshark trace to know what's going on there.

**Comment by jnweiger** [comment]
Thanks @jvillafanez I'll re-run the test and provide icap server logs.

**Comment by jnweiger** [comment]
When trying to upload a file, nothing is logged in the clamav server log, and nothing interesting in the c-icap log files, only
```
cat /var/log/c-icap/access.log
18/Jul/2022:12:33:22 +0000, 172.17.0.3 172.17.0.1 REQMOD avscan?allow204=on&sizelimit=off&mode=simple 400
18/Jul/2022:12:43:08 +0000, 172.17.0.3 172.17.0.1 REQMOD avscan?allow204=on&sizelimit=off&mode=simple 400
18/Jul/2022:12:57:30 +0000, 172.17.0.3 172.17.0.1 REQMOD avscan?allow204=on&sizelimit=off&mode=simple 400
```


Capturing packets...
```
tshark -i docker0
 ** (tshark:29824) 12:55:03.807122 [Main MESSAGE] -- File: "/tmp/wireshark_docker0SMFNP1.pcapng"
    1 0.000000000   172.17.0.1 ? 172.17.0.3   TCP 74 45654 ? 1344 [SYN] Seq=0 Win=64240 Len=0 MSS=1460 SACK_PERM=1 TSval=3582368967 TSecr=0 WS=128
    2 0.000109784   172.17.0.3 ? 172.17.0.1   TCP 74 1344 ? 45654 [SYN, ACK] Seq=0 Ack=1 Win=65160 Len=0 MSS=1460 SACK_PERM=1 TSval=3426825657 TSecr=3582368967 WS=128
    3 0.000221205   172.17.0.1 ? 172.17.0.3   TCP 66 45654 ? 1344 [ACK] Seq=1 Ack=1 Win=64256 Len=0 TSval=3582368967 TSecr=3426825657
    4 0.000832381   172.17.0.1 ? 172.17.0.3   HTTP 7306 PUT / HTTP/1.0 Continuation
    5 0.000875362   172.17.0.3 ? 172.17.0.1   TCP 66 1344 ? 45654 [ACK] Seq=1 Ack=7241 Win=61312 Len=0 TSval=3426825658 TSecr=3582368968
    6 0.000895765   172.17.0.1 ? 172.17.0.3   HTTP 1018 Continuation
    7 0.000903223   172.17.0.3 ? 172.17.0.1   TCP 66 1344 ? 45654 [ACK] Seq=1 Ack=8193 Win=60416 Len=0 TSval=3426825658 TSecr=3582368968
    8 0.000926249   172.17.0.1 ? 172.17.0.3   HTTP 551 Continuation
    9 0.000934026   172.17.0.3 ? 172.17.0.1   TCP 66 1344 ? 45654 [ACK] Seq=1 Ack=8678 Win=60032 Len=0 TSval=3426825658 TSecr=3582368968
   10 0.001119740   172.17.0.3 ? 172.17.0.1   ICAP 175 ICAP/1.0 400 Bad request
   11 0.001154456   172.17.0.1 ? 172.17.0.3   TCP 66 45654 ? 1344 [ACK] Seq=8678 Ack=110 Win=64256 Len=0 TSval=3582368968 TSecr=3426825658
   12 0.001312907   172.17.0.3 ? 172.17.0.1   TCP 66 1344 ? 45654 [RST, ACK] Seq=110 Ack=8678 Win=64128 Len=0 TSval=3426825658 TSecr=3582368968
   13 5.052755391 02:42:4c:f6:e7:9a ? 02:42:ac:11:00:03 ARP 42 Who has 172.17.0.3? Tell 172.17.0.1
   14 5.052729060 02:42:ac:11:00:03 ? 02:42:4c:f6:e7:9a ARP 42 Who has 172.17.0.1? Tell 172.17.0.3
   15 5.052928280 02:42:4c:f6:e7:9a ? 02:42:ac:11:00:03 ARP 42 172.17.0.1 is at 02:42:4c:f6:e7:9a
   16 5.052955811 02:42:ac:11:00:03 ? 02:42:4c:f6:e7:9a ARP 42 172.17.0.3 is at 02:42:ac:11:00:03
^C16 packets captured

**Comment by jvillafanez** [comment]
https://github.com/owncloud/files_antivirus/commit/2e65a7bfd63df69f5d5cde9e3559878a752979bf might be the problem.
According to https://datatracker.ietf.org/doc/html/rfc3507#section-4.4.1 the "req-hdr" and "req-body" should point to the offset where that section start, not the length of the section, and it's weird that we start a header at the byte 35. @s-englert could you confirm?
That could explain why the antivirus chokes: unexpected headers and empty body.

I might have tested it with the master branch, and since that code isn't there yet, that could explain why it worked for me.

**Comment by jvillafanez** [comment]
https://github.com/owncloud/files_antivirus/pull/495 should fix the issue. Note that the PR is against the release-1.1.0 branch.

**Comment by jnweiger** [comment]
Will prepare the next release candidate with this. 
Release is overdue!

**Comment by jnweiger** [comment]
Confirmed fixed in https://github.com/owncloud/files_antivirus/releases/tag/v1.1.0-rc.2

**Comment by jnweiger** [comment]
Argh. That was to simple. Stefan's patch was half right.

We actually have 3 different cases:

* ClamAV works only with `$encapsulated[$type] = \strlen($bodyData);`
* McAfee Webgateway works only with `$encapsulated[$type] = \strlen($data);`
* Fortinet works with both. (seems to not use the Encapsulation header)

Observation: For McAfee Webgateway we send an additional 'Preview' header, that contains exactly `\strlen($data);` so I suspect, McAfee is the special case due to this 'Preview' header.

**Comment by NannaBarz** [comment]
@jvillafanez please work with @jnweiger to get the fix done so we can release soon

**Comment by jnweiger** [comment]
We've merged a workaround #500 - closing here for now.
